lte/gateway/c/oai/tasks/s1ap/s1ap_mme_handlers.c

```diff
@@ -1909,6 +1910,123 @@ int s1ap_mme_handle_ue_context_modification_failure(
 
 //------------------------------------------------------------------------------
 
+int s1ap_mme_handle_handover_request_ack(
+    s1ap_state_t* state, __attribute__((unused)) const sctp_assoc_id_t assoc_id,
+    __attribute__((unused)) const sctp_stream_id_t stream,
+    S1ap_S1AP_PDU_t* pdu) {
+  S1ap_HandoverRequestAcknowledge_t* container = NULL;
+  S1ap_HandoverRequestAcknowledgeIEs_t* ie     = NULL;
+  enb_description_t* source_enb                = NULL;
+  enb_description_t* target_enb                = NULL;
+  hashtable_element_array_t* enb_array         = NULL;
+  uint32_t idx                                 = 0;
+  ue_description_t* ue_ref_p                   = NULL;
+  mme_ue_s1ap_id_t mme_ue_s1ap_id              = INVALID_MME_UE_S1AP_ID;
+  enb_ue_s1ap_id_t tgt_enb_ue_s1ap_id          = INVALID_ENB_UE_S1AP_ID;
+  S1ap_HandoverType_t handover_type            = -1;
+  bstring tgt_src_container                    = {0};
+  imsi64_t imsi64                              = INVALID_IMSI64;
+  s1ap_imsi_map_t* imsi_map                    = get_s1ap_imsi_map();
+
+  OAILOG_FUNC_IN(LOG_S1AP);
+  OAILOG_INFO(LOG_S1AP, "handover request ack received");
+
+  container =
+      &pdu->choice.successfulOutcome.value.choice.HandoverRequestAcknowledge;
+
+  // MME_UE_S1AP_ID: mandatory
+  S1AP_FIND_PROTOCOLIE_BY_ID(
+      S1ap_HandoverRequestAcknowledgeIEs_t, ie, container,
+      S1ap_ProtocolIE_ID_id_MME_UE_S1AP_ID, true);
+  if (ie) {
+    mme_ue_s1ap_id = ie->value.choice.MME_UE_S1AP_ID;
+  } else {
+    OAILOG_FUNC_RETURN(LOG_S1AP, RETURNerror);
+  }
+
+  // eNB_UE_S1AP_ID: mandatory
+  S1AP_FIND_PROTOCOLIE_BY_ID(
+      S1ap_HandoverRequestAcknowledgeIEs_t, ie, container,
+      S1ap_ProtocolIE_ID_id_eNB_UE_S1AP_ID, true);
+  // eNB UE S1AP ID is limited to 24 bits
+  if (ie) {
+    tgt_enb_ue_s1ap_id = (enb_ue_s1ap_id_t)(
+        ie->value.choice.ENB_UE_S1AP_ID & ENB_UE_S1AP_ID_MASK);
+  } else {
+    OAILOG_FUNC_RETURN(LOG_S1AP, RETURNerror);
+  }
+
+  // E-RABAdmittedList: mandatory
+  // Note that the DL and UL transport address and GTP-TEID are optional, and
+  // only used if data forwarding will take place.
+  // TODO: add support for indirect data forwarding
+
+  // Target To Source Transparent Container: mandatory
+  S1AP_FIND_PROTOCOLIE_BY_ID(
+      S1ap_HandoverRequestAcknowledgeIEs_t, ie, container,
+      S1ap_ProtocolIE_ID_id_Target_ToSource_TransparentContainer, true);
+  if (ie) {
+    // note: ownership of tgt_src_container transferred to receiver
+    tgt_src_container = blk2bstr(
+        ie->value.choice.Target_ToSource_TransparentContainer.buf,
+        ie->value.choice.Target_ToSource_TransparentContainer.size);
+  } else {
+    OAILOG_FUNC_RETURN(LOG_S1AP, RETURNerror);
+  }
+
+  // get imsi for logging
+  hashtable_uint64_ts_get(
+      imsi_map->mme_ue_id_imsi_htbl, (const hash_key_t) mme_ue_s1ap_id,
+      &imsi64);
+
+  // get the source eNB -- "source" here refers to the source of handover, not
```

Comment:
Maybe expand on this comment: e.g., `// Retrieve the association ID for the eNB that UE is currently connected (i.e., Source eNB) and pull the Source eNB record from s1ap state using this association ID.`